src/game/common/statemachine.cpp

```diff
@@ -0,0 +1,716 @@
+        return Internal_Set_State(m_defaultStateID);
+    }
+}
+
+void StateMachine::Set_Goal_Object(Object const *obj)
+{
+    if (!m_locked) {
+        Internal_Set_Goal_Object(obj);
+    }
+}
+
+bool StateMachine::Is_Goal_Object_Destroyed() const
+{
+    return m_goalObjectID != OBJECT_UNK && Get_Goal_Object() == nullptr;
+}
+
+void StateMachine::Halt()
+{
+    m_locked = true;
+    m_currentState = nullptr;
+
+#ifdef GAME_DEBUG_STRUCTS
+    if (Get_Wants_Debug_Output()) {
+        captainslog_debug("%d '%s' -- '%s' %x halt()",
+            g_theGameLogic->Get_Frame(),
+            m_owner->Get_Template()->Get_Name().Str(),
+            m_name.Str(),
+            this);
+    }
+#endif
+}
+
+void StateMachine::Internal_Set_Goal_Object(Object const *obj)
+{
+    if (obj) {
+        m_goalObjectID = obj->Get_ID();
+        Internal_Set_Goal_Position(obj->Get_Position());
+    } else {
+        m_goalObjectID = OBJECT_UNK;
+    }
+}
+
+Object *StateMachine::Get_Goal_Object()
+{
+    return g_theGameLogic->Find_Object_By_ID(m_goalObjectID);
+}
+
+Object const *StateMachine::Get_Goal_Object() const
+{
+    return g_theGameLogic->Find_Object_By_ID(m_goalObjectID);
+}
+
+void StateMachine::Set_Goal_Position(Coord3D const *pos)
+{
+    if (!m_locked) {
+        Internal_Set_Goal_Position(pos);
+    }
+}
+
+void StateMachine::Internal_Set_Goal_Position(Coord3D const *pos)
+{
+    if (pos) {
+        m_goalPosition = *pos;
+    }
+}
+
+void StateMachine::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char version = 1;
+    xfer->xferVersion(&version, 1);
+    xfer->xferUnsignedInt(&m_updateFrame);
+    xfer->xferUnsignedInt(&m_defaultStateID);
+    unsigned int id = Get_Current_State_ID();
+    xfer->xferUnsignedInt(&id);
+
+    if (xfer->Get_Mode() == XFER_LOAD) {
+        m_currentState = Internal_Get_State(id);
+    }
+
+    bool b = false;
+    xfer->xferBool(&b);
+
+    if (b) {
+        int count = 0;
+
+        for (auto &iter : m_stateMap) {
+            count++;
+        }
+
+        int count2 = count;
+        xfer->xferInt(&count2);
+        captainslog_dbgassert(count2 == count, "State count mismatch - %d expected, %d read", count, count2);
+
+        for (auto &iter : m_stateMap) {
+            State *state = iter.second;
+            unsigned int state_id = state->Get_ID();
+            xfer->xferUnsignedInt(&state_id);
+
+            captainslog_dbgassert(
+                state_id == state->Get_ID(), "State ID mismatch - %d expected, %d read", state->Get_ID(), state_id);
+
+            if (state == nullptr) {
+                captainslog_dbgassert(state != nullptr, "state was NULL on xfer, trying to heal...");
+                state = Internal_Get_State(m_defaultStateID);
+            }
+
+            state->Xfer_Snapshot(xfer);
+        }
+    } else {
+        if (m_currentState == nullptr) {
+            captainslog_dbgassert(m_currentState != nullptr, "currentState was NULL on xfer, trying to heal...");

```

Comment:
Assert condition can be set 0 here, because `state` was already tested above.

Reply:
fixed.